models/Asset/Service.php

```diff
@@ -493,4 +500,138 @@ public static function getUniqueKey(ElementInterface $element, int $nr = 0): str
+                    $asset->getDao()->deleteFromThumbnailCache($config['thumbnail_config_name']);
+                }
+            }
+
+            if (!$thumbnailConfig) {
+                return null;
+            }
+
+            if ($config['type'] == 'image' && strcasecmp($thumbnailConfig->getFormat(), 'SOURCE') === 0) {
+                $formatOverride = $config['thumbnail_extension'];
+                if (in_array($config['thumbnail_extension'], ['jpg', 'jpeg'])) {
+                    $formatOverride = 'pjpeg';
+                }
+                $thumbnailConfig->setFormat($formatOverride);
+            }
+
+            if ($asset instanceof Asset\Video) {
+                if ($config['type'] == 'video') {
+                    return $asset->getThumbnail($config['thumbnail_name'], [$config['thumbnail_extension']]);
+                } else {
+                    $time = 1;
+                    if (preg_match("|~\-~time\-(\d+)\.|", $config['thumbnail_name'], $matchesThumbs)) {
+                        $time = (int)$matchesThumbs[1];
+                    }
+
+                    return $asset->getImageThumbnail($thumbnailConfig, $time);
+                }
+            } elseif ($asset instanceof Asset\Document) {
+                $page = 1;
+                if (preg_match("|~\-~page\-(\d+)\.|", $config['thumbnail_name'], $matchesThumbs)) {
+                    $page = (int)$matchesThumbs[1];
+                }
+
+                $thumbnailConfig->setName(preg_replace("/\-[\d]+/", '', $thumbnailConfig->getName()));
+                $thumbnailConfig->setName(str_replace('document_', '', $thumbnailConfig->getName()));
+
+                return $asset->getImageThumbnail($thumbnailConfig, $page);
+            } elseif ($asset instanceof Asset\Image) {
+                //check if high res image is called
+
+                preg_match("@([^\@]+)(\@[0-9.]+x)?\.([a-zA-Z]{2,5})@", $config['thumbnail_name'], $matches);
+
+                if (empty($matches) || !isset($matches[1])) {
+                    return null;
+                }
+                if (array_key_exists(2, $matches) && $matches[2]) {
+                    $highResFactor = (float)str_replace(['@', 'x'], '', $matches[2]);
+                    $thumbnailConfig->setHighResolution($highResFactor);
+                }
+
+                // check if a media query thumbnail was requested
+                if (preg_match("#~\-~media\-\-(.*)\-\-query#", $matches[1], $mediaQueryResult)) {
+                    $thumbnailConfig->selectMedia($mediaQueryResult[1]);
+                }
+
+                return $asset->getThumbnail($thumbnailConfig);
+            }
+        }
+
+        return null;
+    }
+
+    #[ArrayShape(['thumbnail_extension' => 'string', 'thumbnail_name' => 'string', 'thumbnail_config_name' => 'string', 'asset_id' => 'string', 'asset_path' => 'string', 'type' => 'image|video'])]
```

Comment:
You should use [PhpDoc array shapes](https://phpstan.org/writing-php-code/phpdoc-types#array-shapes) instead, because PhpStan understands it and PhpStorm, too.

Reply:
Thanks. Changed it accordingly!